syntax = "proto3";

package events;

service PostService {
  rpc CreatePost(CreatePostRequest) returns (PostResponse);
  rpc DeletePost(DeletePostRequest) returns (DeletePostResponse);
  rpc UpdatePost(UpdatePostRequest) returns (PostResponse);
  rpc GetPost(GetPostRequest) returns (PostResponse);
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
  rpc ViewPost(ViewPostRequest) returns (InteractionResponse);
  rpc LikePost(LikePostRequest) returns (InteractionResponse);
  rpc CommentPost(CommentPostRequest) returns (CommentResponse);
  rpc GetComments(GetCommentsRequest) returns (CommentsResponse);
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  string user_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message PostResponse {
  string id = 1;
  string title = 2;
  string description = 3;
  string user_id = 4;
  string created_at = 5;
  string updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message DeletePostRequest {
  string post_id = 1;
  string user_id = 2;
}

message DeletePostResponse { bool success = 1; }

message UpdatePostRequest {
  string post_id = 1;
  optional string title = 2;
  optional string description = 3;
  optional bool is_private = 4;
  repeated string tags = 5;
  string user_id = 6;
}

message GetPostRequest {
  string post_id = 1;
  string user_id = 2;
}

message ListPostsRequest {
  int32 page = 1;
  int32 page_size = 2;
  string user_id = 3;
}

message ListPostsResponse {
  repeated PostResponse posts = 1;
  int32 total = 2;
}

message ViewPostRequest {
  string post_id = 1;
  string user_id = 2;
}

message LikePostRequest {
  string post_id = 1;
  string user_id = 2;
}

message CommentPostRequest {
  string post_id = 1;
  string user_id = 2;
  string content = 3;
}

message GetCommentsRequest {
  string post_id = 1;
  int32 page = 2;
  int32 page_size = 3;
  string user_id = 4;
}

message Comment {
  string id = 1;
  string content = 2;
  string user_id = 3;
  string created_at = 4;
}

message CommentsResponse {
  repeated Comment comments = 1;
  int32 total = 2;
}

message InteractionResponse { bool success = 1; }

message CommentResponse { string comment_id = 1; }